LocalGov Geo upgrade and deployment

andybroomfield opened the pull request:
This test verifies that a custom localgov_geo entity is updated to a geo_entity. Ideally this would be an updateTest, though for simplicitiy this is a functional test that runs the selected update hooks (8002, 8003 and 8004) directly.

andybroomfield committed:
Add test for localgov_geo to geo_entity upgrade

This test verifies that a custom localgov_geo entity is updated to a geo_entity
Ideally this would be an updateTest, though for simplicitiy this is a
functional test that runs the selected update hooks (8002, 8003 and 8004)
directly.

andybroomfield committed:
Add helper class to migrate a display mode and test geo_entity displays

Helper class not part of the update hook so in current form would be run
independently.

andybroomfield commented:
Todo:
- Migrate the form modes
- Migrate custom display modes

Need to decide if we would try to include this in an update hook or leave it as a helper class for other modules that created geos to manage the migration themselves.

andybroomfield committed:
Add test for new geo form mode

Migrate the default form mode and check the add form contains the right values

andybroomfield committed:
Add test for exception and check that the migrated display modes are valids

Check that a form and view display mode that they exist.
Then run the migration.
Check that the display mode is valid before migrating.

finnlewis commented:
Discussing in Merge Monday. 

@andybroomfield is primarily wanting a class to help with the migration to geo_entity.

It would be great to automatically loop through the display modes and migrate those automagically too, in a update hook.

andybroomfield committed:
Extend migration to cover base display modes and add form modes test

This also migrates the appropriate base display view or form mode needed for
a custom display view or form, if it is not yet present (except default which
should always be provided by Drupal).

Also extends the test to cover form mode examples.

Removed the exception tests as they wern't actully performing tests past the
first exception.

andybroomfield commented:
@ekes this is ready for review now and should migrate custom display and form modes from localgov_geo to geo_entity on upgrade.

andybroomfield commented:
And the result
(testing in localgov install)
BHCC Custom parking zones geos can now be migrated.

<img width="882" alt="Screenshot 2023-09-20 at 1 37 27 pm" src="https://github.com/localgovdrupal/localgov_geo/assets/1467480/e7a7f3be-9ddb-4fed-ba5d-fc9014ac9586">

finnlewis commented:
@Adnan-cds this might help with your migration of view modes

andybroomfield committed:
Renumber the last update to 8808 from 8809

ekes commented on modules/localgov_geo_update/src/MigrateDisplayModes.php:
```diff
@@ -0,0 +1,196 @@
+    static::migrateBaseDisplay($type, $name);
+
+    $new_config_name = 'core.entity_' . $type . '_display.geo_entity.' . $bundle . '.' . $name;
+    $new_display_config = $config_factory->getEditable($new_config_name);
```
Where was it that we ended up needing to add UUIDs to entities created this way? I'm guessing we need to do this directly here, I can't remember a better way being bought up.

ekes commented:
Except for the new entities exporting without an UUID it works in a manual test, and automated tests pass.

So fixing UUIDs and then we should merge at the same time as sorting #91 though, as it might also need a by-pass for staging/prod sites with sync config.

andybroomfield commented:
@ekes Drupal should add the UUIDs dynamically itself.
Though this is less of an issue than #91 as if the display mode config entities are deleted and recreated there is no data loss and the config import will take presedence.

andybroomfield commented:
Oh it won't beacuse its wrting the config file directly, but we could change this to create the display mode entity instead and it should get an uuid. and _core hash.

ekes committed:
Fix #91. Don't re-run updates on sites with config in their sync folder.

Add drush deploy function to automatically or manually update content
entities after configuration has been sync'd.

andybroomfield committed:
Generate UUID by saving new display and form mode entities

If a display mode exists, copy the config over.
If not, instead of coping the config, create the display view and form modes
as new entities so they generate UUIDs.

andybroomfield commented:
Ready for re-review.
I've since changed the migration to create display modes as entities if they don't already exist.

ekes commented:
I've got a custom form mode that I made and it's saving with
```
'postal_address' => type => 'localgov_geo_address'
```
so on save from line 128 of MigrateDisplayModes I get a
```
The "localgov_geo_address" plugin does not exist.
```
error.

andybroomfield commented:
@ekes would that be because `localgov_geo_address` should now be `geo_entity_address`?

andybroomfield commented:
@ekes would that be because `localgov_geo_address` should now be `geo_entity_address`?

ekes commented:
That's one of the renames yep: https://git.drupalcode.org/project/geo_entity/-/commit/322e03f6562eec6ce43a5e0908d0fea61554fa2d?page=3#45e68ea6c56efe30722dfdff202b02d70e87b8ef_16_15

ekes commented:
This is the config (`cp config/sync/*localgov_geo* tmp/`) I randomly clicked together to test.
[config.tar.gz](https://github.com/localgovdrupal/localgov_geo/files/12726042/config.tar.gz)

andybroomfield committed:
Rename field plugins on form display from localgov_geo_ to geo_entitiy_

Checks for any type plugins an renames them from localgov_geo versions to
geo_entity equivlents.

Note, the test requries installing the former field from localgov_geo_address
and a stub version for the geo_entity equivilent.

andybroomfield commented:
Do you want to try again @ekes.
Latest should also catch field widgets and rename them, assuming they are all of the type `localgov_geo_` and there replacements are `geo_entitiy_`.

I had some trouble getting a test case for this as localgov_geo no longer ships with the localgov_geo_address widget, and trying to install geo_entitiy_address gives lots of schema errors.

ekes committed:
Merge remote-tracking branch 'origin/fix/91-update-config-export-import' into feature/2.x-update-localgov-geo-tests

ekes committed:
Don't re-use uuid https://github.com/localgovdrupal/localgov_geo/pull/92#issuecomment-1737392972

ekes committed:
Also skip new hook for display and form modes if it's run before.

andybroomfield committed:
Add NULL check in case no config exported

Usuaully this won't be the case (a site upgrade has exported config)
but it was encountered in testing and is a possible edge case, so this
handles that.

andybroomfield commented:
After the merge of #92 into this branch, i'm not seeing the following

>  [notice] Update started: localgov_geo_update_update_8001
>  [warning] Trying to access array offset on value of type bool localgov_geo_update.install:39
>  [error]  array_keys(): Argument #1 ($array) must be of type array, null given 
>  [error]  Update failed: localgov_geo_update_update_8001 

It was because during retesting I deleted all the config sync directory instead of doing a fresh export from Drupal 9.
Its an edge case, but changing
```php
$sync_modules = array_keys($sync['module']);
```
to
```php
$sync_modules = array_keys($sync['module'] ?? []);
```
Allows the update to proceed unhinded.

andybroomfield commented:
Otherwise this update is now precceding as expected.

ekes commented:
I say we get this merged now. But I'll make a MR for another release that we can finalize Monday.

More testers welcome!
